pymc_bart/pgbart.py

```diff
@@ -174,65 +173,55 @@ def astep(self, _):
                 particles = self.resample(particles, normalized_weights)
 
-                # Set the new weight
-                for p in particles[2:]:
-                    p.log_weight = w_t
-
-            for p in particles[2:]:
-                p.log_weight = p.old_likelihood_logp
-
-            _, normalized_weights = self.normalize(particles)
-            # Get the new tree and update
+            normalized_weights = self.normalize(particles)
+            # Get the new particle and associated tree
             self.all_particles[tree_id], new_tree = self.get_particle_tree(
                 particles, normalized_weights
             )
+            # Update the sum of trees
             self.sum_trees = self.sum_trees_noi + new_tree._predict()
+            # To reduce memory usage, we trim the tree
             self.all_trees[tree_id] = new_tree.trim()
 
             if self.tune:
-                self.ssv = SampleSplittingVariable(self.alpha_vec)
+                # Update the splitting variable and the splitting variable sampler
+                if self.iter > self.m:
+                    self.ssv = SampleSplittingVariable(self.alpha_vec)
                 for index in new_tree.get_split_variables():
                     self.alpha_vec[index] += 1
             else:
+                # update the variable inclusion
                 for index in new_tree.get_split_variables():
                     variable_inclusion[index] += 1
 
         if not self.tune:
             self.bart.all_trees.append(self.all_trees)
 
-        stats = {"variable_inclusion": variable_inclusion}
+        stats = {"variable_inclusion": variable_inclusion, "tune": self.tune}
         return self.sum_trees, [stats]
 
     def normalize(self, particles):
-        """Use logsumexp trick to get w_t and softmax to get normalized_weights.
-
-        w_t is the un-normalized weight per particle, we will assign it to the
-        next round of particles, so they all start with the same weight.
+        """
+        Use softmax to get normalized_weights.
         """
         log_w = np.array([p.log_weight for p in particles])
         log_w_max = log_w.max()
         log_w_ = log_w - log_w_max
-        wei = np.exp(log_w_)
-        w_sum = wei.sum()
-        w_t = log_w_max + np.log(w_sum) - self.log_num_particles
-        normalized_weights = wei / w_sum
-        # stabilize weights to avoid assigning exactly zero probability to a particle
-        normalized_weights += 1e-12
-
-        return w_t, normalized_weights
+        wei = np.exp(log_w_) + 1e-12
+        return wei / wei.sum()
 
     def resample(self, particles, normalized_weights):
         """
-        Use systematic resample for all but first two particles
+        Use systematic resample for all but the first particle
 
         Ensure particles are copied only if needed.
         """
-        new_indices = self.systematic(normalized_weights) + 2
+        new_indices = self.systematic(normalized_weights) + 1
```

Comment:
Tbh I don't understand this line either (even before this change) as I thought we sample from the star trajectory as well in each time step of Particle Gibbs. E.g. https://rss.onlinelibrary.wiley.com/doi/full/10.1111/j.1467-9868.2009.00736.x#f2
Notice that the other trajectories can use the star trajectory's nodes.

Reply:
btw, if you have changes in mind that you think can improve the sampler we are totally open to them.

Reply:
I propose to change back to original version of Lakshminarayanan as that has a correct proof. We can try changing the tree to have values at internal node as well so the prediction of each datum is the sum of the value of its path from the root to the leaf. We may want to decrease the prior variance at the node as its depth increases. This would hopefully make the partial tree be more indicative of the full tree's likelihood thus improve mixing.

Reply:
You can propose a PR and we can test if that version provides similar or better performance in terms of fit and mixing.